src/planner.c

```diff
@@ -737,6 +754,126 @@ contain_param(Node *node)
 	return contain_param_exec_walker(node, NULL);
 }
 
+static List *
+fill_missing_groupclause(List *new_groupclause, List *orig_groupclause)
+{
+	if (new_groupclause != NIL)
+	{
+		ListCell *gl;
+		foreach (gl, orig_groupclause)
+		{
+			SortGroupClause *gc = lfirst_node(SortGroupClause, gl);
+
+			if (list_member_ptr(new_groupclause, gc))
+				continue; /* already in list */
+			new_groupclause = lappend(new_groupclause, gc);
+		}
+	}
+	return new_groupclause;
+}
+
+static bool
+check_cagg_view_rte(RangeTblEntry *rte)
+{
+	ContinuousAgg *cagg = NULL;
+	ListCell *rtlc;
+	bool found = false;
+	Query *viewq = rte->subquery;
+	Assert(rte->rtekind == RTE_SUBQUERY);
+
+	if (list_length(viewq->rtable) != 3) // a view has 3 entries
+	{
+		return false;
+	}
+
+	// should cache this information for cont. aggregates
+	foreach (rtlc, viewq->rtable)
+	{
+		RangeTblEntry *rte = (RangeTblEntry *) lfirst(rtlc);
+		char *schema;
+		char *table;
+
+		if (!OidIsValid(rte->relid))
+			break;
+
+		schema = get_namespace_name(get_rel_namespace(rte->relid));
+		table = get_rel_name(rte->relid);
+		if ((cagg = ts_continuous_agg_find_by_view_name(schema, table)) != NULL)
+			found = true;
+	}
+	return found;
+}
+
+/* Note that it modifies the passed in Query
+* select * from (select a, b, max(c), min(d) from ...
+				 group by a, b)
+  order by b;
+* is transformed as
+* SELECT * from (select a, b, max(c), min(d) from ..
+*                 group by B, A  <------ note the change in order here
+*              )
+*  order by b;
+*  we pushdown only if order by is a subset of group-by
+* restrict this to continuous aggregates
+*/
+static void
+cagg_reorder_groupby(Query *query)
+{
+	ListCell *lc, *rtlc;
+	int rtno = 0;
+	List *outer_tlist = query->targetList;
+	List *outer_sortcl = query->sortClause;
+
+	foreach (rtlc, query->rtable)
+	{
+		RangeTblEntry *rte = (RangeTblEntry *) lfirst(rtlc);
+		bool not_found = true;
+		rtno++;
+		if (rte->rtekind == RTE_SUBQUERY)
+		{
+			if (outer_sortcl && rte->subquery->groupClause && rte->subquery->sortClause == NIL &&
+				check_cagg_view_rte(rte))
+			{
+				List *new_groupclause = NIL;
+				foreach (lc, outer_sortcl)
+				{
+					SortGroupClause *outer_sc = (SortGroupClause *) lfirst(lc);
+					TargetEntry *outer_tle = get_sortgroupclause_tle(outer_sc, outer_tlist);
+					not_found = true;
+					if (IsA(outer_tle->expr, Var) && (((Var *) outer_tle->expr)->varno == rtno))
```

Comment:
nit: to reduce the nesting level and increase readabilitiy. can we change from if(condition) { if (condition) ... not_found = false} to if(!condition) break; ... if(!condition) break ... not_found = false;

Reply:
or break out into a separate function